### PR TITLE
Add Animation Record UI

### LINKED ISSUE #345: Start headmover with default headmode
# Summary

## Expected behavior

When starting the motion, or head is still disabled.
We should also rise the head.
Thus, start the headmover with motion_standalone.
The teleop_keyboard should also set headmodes (if not already)

## Steps to Reproduce

1.
2.
3.

### LINKED ISSUE #359: Head mover dies
The head mover dies with this traceback if the motors start a little bit late:

```
[taskset-25] [INFO] [dynamic_kick]: Initialized Kick and waiting for actions (KickNode() at /home/bitbots/colcon_ws/src/bitbots_motion/bitbots_dynamic_kick/src/kick_node.cpp:124)
[move_head-20] Stack trace (most recent call last):
[move_head-20] #12   Object "/usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2", at 0xffffffffffffffff, in 
[move_head-20] #11   Object "/home/bitbots/colcon_ws/build/bitbots_head_mover/move_head", at 0x55ae3a8a0784, in _start
[move_head-20] #10   Source "../csu/libc-start.c", line 392, in __libc_start_main_impl [0x7f164f629e3f]
[move_head-20] #9    Source "../sysdeps/nptl/libc_start_call_main.h", line 58, in __libc_start_call_main [0x7f164f629d8f]
[move_head-20] #8    Source "/home/bitbots/colcon_ws/src/bitbots_motion/bitbots_head_mover/src/move_head.cpp", line 796, in main [0x55ae3a8ae76e]
[move_head-20]         793:   rclcpp::experimental::executors::EventsExecutor exec;
[move_head-20]         794:   auto head_mover = std::make_shared<move_head::HeadMover>();
[move_head-20]         795:   exec.add_node(head_mover->get_node());
[move_head-20]       > 796:   exec.spin();
[move_head-20]         797:   rclcpp::shutdown();
[move_head-20]         798: 
[move_head-20]         799:   return 0;
[move_head-20] #7    Object "/opt/ros/iron/lib/librclcpp.so", at 0x7f164ff151cf, in rclcpp::experimental::executors::EventsExecutor::spin()
[move_head-20] #6    Object "/opt/ros/iron/lib/librclcpp.so", at 0x7f164ff1aaf6, in rclcpp::experimental::TimersManager::execute_ready_timer(rclcpp::TimerBase const*)
[move_head-20] #5    Source "/opt/ros/iron/include/rclcpp/rclcpp/timer.hpp", line 273, in execute_callback [0x55ae3a97ab2c]
[move_head-20]         270:   execute_callback() override
[move_head-20]         271:   {
[move_head-20]         272:     TRACEPOINT(callback_start, reinterpret_cast<const void *>(&callback_), false);
[move_head-20]       > 273:     execute_callback_delegate<>();
[move_head-20]         274:     TRACEPOINT(callback_end, reinterpret_cast<const void *>(&callback_));
[move_head-20]         275:   }
[move_head-20] #4    Source "/opt/ros/iron/include/rclcpp/rclcpp/timer.hpp", line 287, in execute_callback_delegate<> [0x55ae3a980b4f]
[move_head-20]         284:   void
[move_head-20]         285:   execute_callback_delegate()
[move_head-20]         286:   {
[move_head-20]       > 287:     callback_();
[move_head-20]         288:   }
[move_head-20]         289: 
[move_head-20]         290:   template<
[move_head-20] #3    Source "/home/bitbots/colcon_ws/src/bitbots_motion/bitbots_head_mover/src/move_head.cpp", line 184, in operator() [0x55ae3a8d4088]
[move_head-20]         181:         std::bind(&HeadMover::handle_accepted, this, std::placeholders::_1));
[move_head-20]         182: 
[move_head-20]         183:     // Initialize timer for main loop
[move_head-20]       > 184:     timer_ = rclcpp::create_timer(node_, node_->get_clock(), 10ms, [this] { behave(); });
[move_head-20]         185:   }
[move_head-20]         186: 
[move_head-20]         187:   /**
[move_head-20] #2    Source "/home/bitbots/colcon_ws/src/bitbots_motion/bitbots_head_mover/src/move_head.cpp", line 767, in behave [0x55ae3a8db499]
[move_head-20]         764:       }
[move_head-20]         765: 
[move_head-20]         766:       // Get the current head position
[move_head-20]       > 767:       std::pair<double, double> head_position = get_head_position();
[move_head-20]         768: 
[move_head-20]         769:       // Select the closest keypoint in the search pattern as a starting point
[move_head-20]         770:       index_ = get_near_pattern_position(pattern_, head_position.first, head_position.second);
[move_head-20] #1    Source "/home/bitbots/colcon_ws/src/bitbots_motion/bitbots_head_mover/src/move_head.cpp", line 462, in get_head_position [0x55ae3a8d9397]
[move_head-20]         459:     double head_tilt = 0.0;
[move_head-20]         460: 
[move_head-20]         461:     // Iterate over all joints and find the head pan and tilt joints
[move_head-20]       > 462:     for (size_t i = 0; i < current_joint_state_->name.size(); i++) {
[move_head-20]         463:       if (current_joint_state_->name[i] == "HeadPan") {
[move_head-20]         464:         head_pan = current_joint_state_->position[i];
[move_head-20]         465:       } else if (current_joint_state_->name[i] == "HeadTilt") {
[tts-17] Playing WAVE 'stdin' : Signed 16 bit Little Endian, Rate 22050 Hz, Mono
[move_head-20] #0    Source "/usr/include/c++/11/bits/stl_vector.h", line 919, in size [0x55ae3a8dcdd0]
[move_head-20]         916:       /**  Returns the number of elements in the %vector.  */
[move_head-20]         917:       size_type
[move_head-20]         918:       size() const _GLIBCXX_NOEXCEPT
[move_head-20]       > 919:       { return size_type(this->_M_impl._M_finish - this->_M_impl._M_start); }
[move_head-20]         920: 
[move_head-20]         921:       /**  Returns the size() of the largest possible %vector.  */
[move_head-20]         922:       size_type
[move_head-20] Segmentation fault (Address not mapped to object [0x30])
```

This should not happen!

### PR DESCRIPTION
**Pull request by Flova**
# Changes
- Ports the old record animation rqt tool from ROS1 to ROS2
- Heavily rewrite the animation record GUI, fixing bugs and cleaning / simplifying the code
- Redo the handling of animations in the HCM. Non HCM animations need to call a service to enter the HCM animation state, instead of sending a specially formatted message on the animation joint state topic. This simplifies rejections of requests, the joint mutex itself, as well as the animation joint command message.
- Also redo the handling of the recording mode in the HCM. It is activated via a service if the recording GUI plays an animation and prohibits falling reactions etc.. It is deactivated once the GUI is closed.
- Dynamic update/addition of animations during runtime. This allows for a way less hacky way of playing work in progress animations from the GUI. Previously they were copied via ssh onto the ROS master uri host, guessing the folder structure. Not it is realized via a ROS service that accepts a json serialized animation. The service is used because a client might want to wait for the animation to be added before it is executed and a service can be blocking.
- Add the ability to partially play animations for debugging purposes (mainly used by the GUI to step through the frames manually). This allows for more accurate debugging as the debugged animation is played using the animation server and not a different animation runtime inside the GUI which handles certain scenarios different and is overall code duplication.
- Fix misuse of the interactive shell command `exit()`, which resulted in freezes. `sys.exit()` or proper exceptions are used instead.
- Addition of a robot binding script. It sets all the environment variables to enable seamless ROS communication with a robot given the robot's name. It also restarts the local ros2 daemon, which is a typical source of errors when switching domain IDs.
- Start head mover with motion instead of behavior. This makes semantically more sense and it is also cleaner dependency wise. (Closes #345 )
- Fix uninitialized memory error in head mover
- Renaming some minor things
- Fix head mover death (closes #359)

### CODE REVIEW

File: bitbots_msgs/msg/Animation.msg [hunk 1]
```diff
@@ -1,16 +1,4 @@
-std_msgs/Header header
+# Is this animation coming from the hardware control manager and can be executed duing all robot states
```
Comment (timonegk) [inline bitbots_msgs/msg/Animation.msg:1]:
```suggestion
# Is this animation coming from the humanoid control module and can be executed duing all robot states
```

File: bitbots_msgs/action/PlayAnimation.action [hunk 2]
```diff
@@ -3,11 +3,20 @@
 # Name of the animation
 string animation
 
-# If the animation call comes from the hardware control manager, 
+# If the animation call comes from the hardware control manager,
```
Comment (timonegk) [inline bitbots_msgs/action/PlayAnimation.action:6]:
```suggestion
# If the animation call comes from the humanoid control module,
```

File: bitbots_motion/bitbots_head_mover/src/move_head.cpp [hunk 3]
```diff
@@ -56,7 +56,7 @@ class HeadMover {
   std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
 
   // Declare variables
-  uint head_mode_;
+  uint head_mode_ = 7;
```
Comment (timonegk) [inline bitbots_motion/bitbots_head_mover/src/move_head.cpp:59]: Please avoid magic values.
```suggestion
  uint head_mode_ = humanoid_league_msgs::msg::HeadMode::LOOK_FORWARD;
```

Comment (Flova) [reply bitbots_motion/bitbots_head_mover/src/move_head.cpp:59]: Oh yeah this was by accident. I totally agree with you.

File: bitbots_msgs/action/PlayAnimation.action [hunk 4]
```diff
@@ -3,11 +3,20 @@
...
+# If the animation call comes from the humanoid control module,
 # it should be played directly, interrupting other animations and walking
 # It's probably falling or standing up
 bool hcm
 
+# Debugging controls
+bool bounds # If true, only a specific range of keyframes will be played
+# Keyframe where we want to start (use 0 as the start of the animation)
+uint8 start
+# Keyframe where we want to end (use len(animation) as the end of the animation)
```
Comment (timonegk) [inline bitbots_msgs/action/PlayAnimation.action:15]: Is it really `len(animation)` or should it be `len(animation) - 1`? If it should be the latter, changes in `animation_node.py` are also necessary.

**Review by timonegk — APPROVED** [review]
Apart from some comments, the changes look good architecture-wise. Assuming you tested the code, I think this can be merged.
I didn't test or read the GUI code as it is not critical.

File: bitbots_msgs/action/PlayAnimation.action (hunk 4 above)
Comment (Flova) [reply bitbots_msgs/action/PlayAnimation.action:15]: It is len(animation) like the [start:end] syntax in numpy

**Pull request merged**